SCClassLibrary/Common/Math/Integer.sc

```diff
@@ -15,13 +15,17 @@ Integer : SimpleNumber {
 	odd { ^this.bitAnd(1) == 1 }
 
 	xrand { arg exclude=0;
-		^(exclude + (this - 1).rand + 1) % this;
+		var res = this.rand;
+		if ( res == exclude ) { res = this };
+		^res;
 	}
+
 	xrand2 { arg exclude=0;
-		var res;
-		res = (2 * this).rand - this;
-		if (res == exclude, { ^this },{ ^res });
+		var res = this.rand2;
+		if ( res == exclude ) { res = [this, this.neg].choose };
+		^res;
```

Comment:
this algorithm unfortunately doesn't produce an even distribution; you can use `Collection:-plotHisto` to quickly check if your output makes sense. Note that `rand2`'s range is `[-this, this]` (inclusive) while `rand` is `[0, this)` (exclusive).